chiron/integrators.py

```diff
@@ -137,10 +137,30 @@ def run(
         b = jnp.sqrt(1 - jnp.exp(-2 * collision_rate_unitless * stepsize_unitless))
 
         # Initialize velocities
-        if self.velocities is None:
-            v0 = sigma_v * random.normal(key, x0.shape)
-        else:
-            v0 = self.velocities.value_in_unit_system(unit.md_unit_system)
+        if self.reinitialize_velocities:
+            # v0 = sigma_v * random.normal(key, x0.shape)
+            from .utils import initialize_velocities
+
+            sampler_state.velocities = initialize_velocities(
+                temperature, potential.topology, key
+            )
+
+        elif sampler_state._velocities is None:
+            # v0 = sigma_v * random.normal(key, x0.shape)
+            from .utils import initialize_velocities
+
+            sampler_state.velocities = initialize_velocities(
+                temperature, potential.topology, key
+            )
+        elif sampler_state._velocities.shape[0] != sampler_state.x0.shape[0]:
+            from .utils import initialize_velocities
+
+            sampler_state.velocities = initialize_velocities(
+                temperature, potential.topology, key
+            )
+
+        # extract the velocities from the sampler state
+        v0 = sampler_state.velocities
```

Comment:
I think those are the 3 parts that are going to need the most attention. 

Right now, the pair computation (currently limited just to LJ) and neighborlist functions are already coded up with the appropriate functions jitted for performance. 

I had done some "offline" (i.e., outside of this PR) benchmarking of the langevin integrator, looking at different approaches to how we jit the routine (e.g., there are a few consistent patterns that can be coded up to minimize overhead of jit).  Once this is all merged, I was going to create a few benchmarks of different approaches in an issue so we can figure out, discuss, brainstorm, etc. the most efficient implementation.    Preliminarily, it seems that jax.numpy itself is already quite efficient even without jitting the integrator,  which I suspect is because the amount of "work" is much much less compared to, say computing the potential between all the particles.

Reply:
I've added this discussion to issue #34 .